test/SILOptimizer/devirtualize_protocol_composition_two_stores.sil

```diff
@@ -0,0 +1,114 @@
+// RUN: %target-build-swift %s -O -wmo -emit-sil | %FileCheck %s
+
+sil_stage canonical
+
+import Builtin
+import Swift
+import SwiftShims
+
+public class A {
+  @objc deinit
+  init()
+}
+
+public protocol P {
+  func foo() -> Int
+}
+
+public class B : A, P {
+  public func foo() -> Int
+  @objc deinit
+  override init()
+}
+
+public class C : A, P {
+  public func foo() -> Int
+  @objc deinit
+  override init()
+}
+
+func imp(x: A & P, y: A & P) -> Int
+
+public func test(x: B, y: C) -> Int
+
+// CHECK-LABEL: sil shared [noinline] @{{.*}}impl{{.*}}
+// In the optimization pass we look for uses of alloc_stack (aka %5).
+// This test makes sure that we look at the correct uses with the
+// correct dominance order (store before apply before dealloc).
+// This function will be passed arguments of type B and C for arguments
+// %0 and %1 respectively. We want to make sure that we call B's foo method
+// and not C's foo method.
+sil hidden [noinline] @impl : $@convention(thin) (@guaranteed A & P, @guaranteed A & P) -> Int {
+bb0(%0 : $A & P, %1 : $A & P):
+  %2 = open_existential_ref %0 : $A & P to $@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  %3 = unchecked_ref_cast %1 : $A & P to $@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+
+  %5 = alloc_stack $@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  store %2 to %5 : $*@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  
+  // We want to make sure that we picked up on the FIRST store and not the second one.
+  // class C's foo method is named "bar" whereas class B's foo method is named "foo".
+  // We want to make sure that we call a function named "foo" not "bar".
+  // CHECK: [[FN:%[0-9]+]] = function_ref @${{.*}}foo{{.*}} : $@convention(thin) () -> Int
+  %7 = witness_method $@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P, #P.foo!1 : <Self where Self : P> (Self) -> () -> Int, %2 : $@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P : $@convention(witness_method: P) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> Int
+  // CHECK: apply [[FN]]
+  %8 = apply %7<@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P>(%5) : $@convention(witness_method: P) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> Int
+  
+  store %3 to %5 : $*@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  dealloc_stack %5 : $*@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  return %8 : $Int
+} // end sil function 'impl'
```

Comment:
It's generally good habit to also mark the function end with `CHECK-LABEL`. It catches any mistakes in the next CHECK-LABEL and cuts down on confusing `lit` errors.

Reply:
Good to know. Thanks.